### ISSUE TITLE
Load Profile window doesn't show the password box - impossible to login

### ISSUE THREAD

**Issue by yurivict** [qTox/qTox#3530]
It used to ask the password. At some point password box stopped being displayed during startup. Pressing "Load" causes the "Wrong password" window.

This is a severe usability bug because the user is unable to login at all.
##### Brief Description

OS: FreeBSD
qTox version: 1.4.1
toxcore:  0.151112
Qt:  5.5.1

**Comment by zetok** [comment]
Works for me with both newest qTox (f970bb5eaca45c5500b1723ff82116e5ffb0a878) and 1.4.0.
![works](https://cloud.githubusercontent.com/assets/3148759/16943958/46db0ecc-4d96-11e6-89c2-16b3137038ea.png)

This might be a duplicate of #2587. If so, close this one, and provide relevant info in #2587.

Could you see if renaming `~/.config/tox/qtox.ini` before starting qTox to something else would change anything? And if so, would you mind providing this corrupted settings file as a test case?

**Comment by zetok** [comment]
Can anyone else reproduce?

**Comment by yurivict** [comment]
Renaming ~/.config/tox/qtox.ini didn't help.

**Comment by sudden6** [comment]
could be a corrupted tox profile or database, what does the log say?

**Comment by zetok** [comment]
@sudden6 err? those shouldn't affect the UI..

**Comment by yurivict** [comment]
I copied the ~/.config/tox folder to my main FreeBSD machine and reproduced the problem. This account is just a test account - nothing valuable there, so I am attaching the whole directory zip file.

[tox-problem.zip](https://github.com/tux3/qTox/files/372655/tox-problem.zip)

**Comment by antis81** [comment]
> err? those shouldn't affect the UI..

Obviously they do! :) Looks like the tox profile is corrupt indeed (0 byte size).

@yurivict Posting the logfile and/or coredump would help with such cases.

**Comment by yurivict** [comment]
It looks like the profile update isn't handled right.

Profile is supposed to be an SqLite database, that is super-reliable.
Program isn't supposed to save the copy of the old file under the different name and then write a new one and delete the old one. This process is error-prone.

**Comment by antis81** [comment]
@yurivict Please post the log file. It should tell about that.

**Comment by antis81** [comment]
To me this sounds like a typical migration problem. In this case, qTox indeed should create a temporary copy (like e.g. "profile.db.<old-version>"), before migration takes place. This file should be deleted, as soon as migration succeeds.

**Comment by yurivict** [comment]
Migration from where to where?

**Comment by antis81** [comment]
> Migration from where to where?

Old profile -> new profile. This is just an assumption, as I'm not too deep into that topic.

**Comment by yurivict** [comment]
There should be no old and new profile. There is only one profile. It is error-prone to have such logic.

**Comment by antis81** [comment]
It's about migration, right? So yes, there's an old and a new format. Sorry, of course there's not an old and new profile, but the format changed.

**Comment by yurivict** [comment]
This profile was created by the same version of qTox, so there is no migration.

**Comment by antis81** [comment]
> This profile was created by the same version of qTox.

Huh? So you created a (the first) profile with the current qTox version and it doesn't load? Is that what you mean? Btw. the default profile doesn't have a password and completely starts without password prompt.

**Comment by yurivict** [comment]
This profile was created on the same machine with the same executable an hour eariler, so no migration should have taken place. I think qTox saves the new version of profile by renaming the file and writing a new copy, which is a completely wrong way to do this.

**Comment by zetok** [comment]
> Profile is supposed to be an SqLite database

No, it's not.

**Comment by yurivict** [comment]
If it isn't in SqLite, the next safest thing is QSaveFile.

**Comment by zetok** [comment]
@yurivict do you know what could have caused profile corruption?

**Comment by yurivict** [comment]
On the laptop network card driver was faulty and laptop crashed several times. This eventually caused this problem with the qTox profile. This is why program shouldn't ever delete and write again the sensitive files. They should ideally be updated in-place. Or the backup file should be in place with an automatic recovery process.

**Comment by sudden6** [comment]
qTox already uses QSaveFile, which creates a new file and then renames it to overwrite the old file. This is already an atomic operation on the program level and pretty much guarantees that the data is actually written. UNTIL your OS crashes!

 If your OS uses a disk cache (99% yes) the data is only really written to disk in specific intervals. If your OS crashes in between you're out of luck, not only for qTox, but also for every other program having data in the disk cache.

**Comment by yurivict** [comment]
> data is only really written to disk in specific intervals.

Unless sync(2) is called, in which case the cache is written to disk immediately.

The problem is that the profile is rewritten as a whole. This is why it would have been better if the profile was an SQLite db. They ensure that in-place updates are as reliable as possible. It would have had at least the previous copy of the data.

An alternative is to first write the backup copy of the profile, to sync, then to write the profile file and sync again. During load, if the main profile file is invalid, just use the backup copy instead.

**Comment by sudden6** [comment]
> The problem is that the profile is rewritten as a whole. This is why it would have been better if the profile was an SQLite db. They ensure that in-place updates are as reliable as possible. It would have had at least the previous copy of the data.

The profile can't be an SQLite db because of compatibility with other Tox clients.

> Unless sync(2) is called, in which case the cache is written to disk immediately.

Yes, but calling sync(2) all the time will slow everything down and cause lot's of unnecessary writes.

Also worth to mention, because this were attempts to fix similiar stuff, but failed for various reasons:

https://github.com/tux3/qTox/pull/3082
https://github.com/tux3/qTox/pull/3192

**Comment by sudden6** [comment]
Reducing priority since the initial issue was a OS crash and qTox can't really prevent that.

**Comment by anthonybilinski** [comment]
So to summarize:
1) Prompting for password from the login screen is based on the tox profile save being encrypted
2) The file format is cross-client, so we can't easily change it to sqlite.
3) We use QSaveFile with writes to a temporary file, which we flush, and only on success swap with the file destination.

I've read that FreeBSD like Mac OS does a slightly less aggresive sync by just pushing data to the drive cache on fsync by default, so the issue might be more likely to be seen there. But I don't think there's anything really actionable by us here.

Maybe in the distant future we could store user settings, global settings, and the tox profile all in our current chat history DB and just offer an option to export the tox profile from there to be imported by other clients.

Closing as there's nothing immediately actionable.
